spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TempDirExtension.java

```diff
@@ -0,0 +1,173 @@
+package org.spockframework.runtime.extension.builtin;
+
+import org.spockframework.runtime.InvalidSpecException;
+import org.spockframework.runtime.extension.AbstractMethodInterceptor;
+import org.spockframework.runtime.extension.IAnnotationDrivenExtension;
+import org.spockframework.runtime.extension.IMethodInvocation;
+import org.spockframework.runtime.model.FieldInfo;
+import org.spockframework.runtime.model.SpecInfo;
+import org.spockframework.util.WrappedException;
+import spock.lang.TempDir;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.nio.file.*;
+import java.nio.file.attribute.BasicFileAttributes;
+
+import static java.nio.file.FileVisitResult.CONTINUE;
+
+/**
+ * @author: dqyuan
+ * @date: 2020/07/12
+ */
+public class TempDirExtension implements IAnnotationDrivenExtension<TempDir> {
+
+  private static final String TEMP_DIR_PREFIX = "spock";
+
+  @Override
+  public void visitFieldAnnotation(TempDir annotation, FieldInfo field) {
+    Class<?> fieldType = field.getType();
+    if (!File.class.equals(fieldType) && !Path.class.equals(fieldType)) {
+      throw new InvalidSpecException("@TempDir can only be used on File field or Path field");
+    }
+    DirInterceptor interceptor = field.isShared()?
+      new SharedInterceptor(fieldType, field.getReflection()):
+      new FeatureInterceptor(fieldType, field.getReflection());
+    interceptor.install(field.getParent());
+  }
+
+  private static Path generateTempDir() {
+    try {
+      return Files.createTempDirectory(TEMP_DIR_PREFIX);
+    } catch (IOException e) {
+      throw new WrappedException(e);
+    }
+  }
+
+  private abstract static class DirInterceptor extends AbstractMethodInterceptor {
+    private final Class<?> fieldType;
+    private final Field reflection;
+    private Path tempPath;
+
+    DirInterceptor(Class<?> fieldType, Field reflection) {
+      this.fieldType = fieldType;
+      this.reflection = reflection;
+    }
+
+    void setUp(IMethodInvocation invocation)  throws Throwable {
+      tempPath = generateTempDir();
+      reflection.setAccessible(true);
+      reflection.set(invocation.getInstance(), fieldType == Path.class? tempPath: tempPath.toFile());
+      invocation.proceed();
+    }
+
+    void destroy(IMethodInvocation invocation) throws Throwable {
+      try {
+        invocation.proceed();
+      } finally {
+        deleteTempDir();
+      }
+    }
+
+    private void deleteTempDir() throws IOException {
```

Comment:
You can just use `ResourceGroovyMethods.deleteDir(tempPath.toFile())`, Groovy implements the recursive logic already.
If it cannot delete non-writable content, you could check the return value, make a pass that just sets everything to writable if possible and then use the method again.